lib/learn_to_rank/data_pipeline/embed_features.rb

```diff
@@ -1,33 +1,33 @@
 module LearnToRank::DataPipeline
   class EmbedFeatures
     # EmbedFeatures takes a set of relevancy judgements and add features to them.
-    # INPUT: [{ query: "a1", "id": "123456", score: 1 }]
-    # OUTPUT: [{ query: "a1", "id": "123456", score: 1,
-    #            view_count: 2, description_score: 2, title_score: 4 ... }]
+    # INPUT: enumerator of { query: "a1", "id": "123456", score: 1 }
+    # OUTPUT: lazy enumerator of { query: "a1", "id": "123456", score: 1, view_count: 2, description_score: 2, title_score: 4 ... }
     def initialize(judgements)
       @judgements = judgements
-      @cached_queries = {}
     end
 
     def augmented_judgements
-      count = Float(judgements.count)
-      res = judgements.compact.map.with_index do |judgement, i|
-        logger.info "#{i}/#{count}: #{(i / count) * 100}%"
-        feats = features(judgement)
+      cached_queries = {}
+      augmented = judgements.lazy.map do |judgement|
+        next nil unless judgement
+
+        feats = features(judgement, cached_queries)
         next nil unless feats
 
+        cached_queries = {} if cached_queries.keys.count > 25
```

Comment:
Are judgements already sorted by query? Just wondering since it'd help the caching here.

Reply:
They should be, as the bigquery groups on `searchTerm`